File: tests/test_thumbnail.py
"""Tests for thumbnail generation and caching."""

from pathlib import Path

from PIL import Image

from src.core.thumbnail import ThumbnailCache, _cache_key


def test_cache_key_deterministic(sample_jpeg):
    key1 = _cache_key(sample_jpeg)
    key2 = _cache_key(sample_jpeg)
    assert key1 == key2
    assert len(key1) == 64


def test_cache_key_changes_with_mtime(sample_jpeg):
    key1 = _cache_key(sample_jpeg)
    sample_jpeg.touch()
    key2 = _cache_key(sample_jpeg)
    assert key1 != key2


def test_generate_thumbnail(sample_jpeg, tmp_path):
    cache = ThumbnailCache(cache_dir=tmp_path / "cache")
    thumb = cache.get_thumbnail(sample_jpeg)
    assert isinstance(thumb, Image.Image)
    assert max(thumb.size) <= 256


def test_cache_hit(sample_jpeg, tmp_path):
    cache = ThumbnailCache(cache_dir=tmp_path / "cache")
    thumb1 = cache.get_thumbnail(sample_jpeg)
    thumb2 = cache.get_thumbnail(sample_jpeg)
    assert thumb1.size == thumb2.size
    key = _cache_key(sample_jpeg)
    assert (tmp_path / "cache" / f"{key}.png").exists()


def test_nonexistent_file(tmp_path):
    cache = ThumbnailCache(cache_dir=tmp_path / "cache")
    result = cache.get_thumbnail(tmp_path / "nope.jpg")
    assert result is None


def test_large_image_thumbnail(tmp_image_dir, tmp_path):
    path = tmp_image_dir / "big.jpg"
    img = Image.new("RGB", (4000, 3000), color=(100, 100, 100))
    img.save(path, "JPEG")
    cache = ThumbnailCache(cache_dir=tmp_path / "cache")
    thumb = cache.get_thumbnail(path)
    assert max(thumb.size) <= 256
    assert thumb.size[0] > 0 and thumb.size[1] > 0

File: tests/test_thumbnail.rs
use std::path::PathBuf;
use image::{DynamicImage, RgbImage};
use omarchy_imageview::core::thumbnail::{ThumbnailCache, cache_key};

fn create_test_jpeg(dir: &std::path::Path, name: &str, w: u32, h: u32) -> PathBuf {
    let path = dir.join(name);
    let img = DynamicImage::ImageRgb8(RgbImage::new(w, h));
    img.save(&path).unwrap();
    path
}

#[test]
fn test_cache_key_deterministic() {
    let dir = tempfile::tempdir().unwrap();
    let path = create_test_jpeg(dir.path(), "test.jpg", 100, 80);
    let k1 = cache_key(&path).unwrap();
    let k2 = cache_key(&path).unwrap();
    assert_eq!(k1, k2);
    assert_eq!(k1.len(), 64);
}

#[test]
fn test_generate_thumbnail() {
    let dir = tempfile::tempdir().unwrap();
    let cache_dir = tempfile::tempdir().unwrap();
    let path = create_test_jpeg(dir.path(), "test.jpg", 100, 80);
    let cache = ThumbnailCache::new(Some(cache_dir.path().to_path_buf()));
    let thumb = cache.get_thumbnail(&path).unwrap();
    assert!(thumb.width() <= 256 && thumb.height() <= 256);
}

#[test]
fn test_cache_hit() {
    let dir = tempfile::tempdir().unwrap();
    let cache_dir = tempfile::tempdir().unwrap();
    let path = create_test_jpeg(dir.path(), "test.jpg", 100, 80);
    let cache = ThumbnailCache::new(Some(cache_dir.path().to_path_buf()));
    let _ = cache.get_thumbnail(&path);
    let key = cache_key(&path).unwrap();
    assert!(cache_dir.path().join(format!("{}.png", key)).exists());
}

#[test]
fn test_nonexistent() {
    let cache_dir = tempfile::tempdir().unwrap();
    let cache = ThumbnailCache::new(Some(cache_dir.path().to_path_buf()));
    assert!(cache.get_thumbnail(std::path::Path::new("/nonexistent.jpg")).is_none());
}

#[test]
fn test_large_image_scaled() {
    let dir = tempfile::tempdir().unwrap();
    let cache_dir = tempfile::tempdir().unwrap();
    let path = create_test_jpeg(dir.path(), "big.jpg", 4000, 3000);
    let cache = ThumbnailCache::new(Some(cache_dir.path().to_path_buf()));
    let thumb = cache.get_thumbnail(&path).unwrap();
    assert!(thumb.width() <= 256);
    assert!(thumb.height() <= 256);
}
